integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITConcurrentStreamAppendReplaceTest.java

```diff
@@ -0,0 +1,471 @@
+
+      final Interval minute3 = Intervals.of("2000-01-01T01:03:00Z/2000-01-01T01:04:00Z");
+      long rowsForMinute3 = generateData(minute3, streamEventWriter);
+
+      // Wait for data to be ingested for all the minutes
+      ensureRowCount(rowsForMinute1 + rowsForMinute2 + rowsForMinute3);
+
+      ensureSegmentsLoaded();
+
+      // 2 segments for each minute, total 6
+      ensureSegmentsCount(6);
+      printTaskStatuses();
+
+      // Trigger auto compaction
+      submitAndVerifyCompactionConfig(null);
+      compactionResource.forceTriggerAutoCompaction();
+
+      // Verify that all the segments are now compacted
+      ensureRowCount(rowsForMinute1 + rowsForMinute2 + rowsForMinute3);
+      ensureSegmentsLoaded();
+      ensureSegmentsCount(3);
+      verifyCompactedIntervals(minute1, minute2, minute3);
+      printTaskStatuses();
+
+      // Concurrent compaction with configured segment granularity
+      for (int i = 0; i < 5; i++) {
+        rowsForMinute1 += generateData(minute1, streamEventWriter);
+        rowsForMinute2 += generateData(minute2, streamEventWriter);
+        rowsForMinute3 += generateData(minute3, streamEventWriter);
+
+        //ensureNoLockContention();
+        compactionResource.forceTriggerAutoCompaction();
+        ensureRowCount(rowsForMinute1 + rowsForMinute2 + rowsForMinute3);
+        printTaskStatuses();
+        checkAndSetConcurrentLocks();
+      }
+
+      // Verify the state with minute granularity
+      ensureSegmentsCount(3);
+      ensureSegmentsLoaded();
+      ensureRowCount(rowsForMinute1 + rowsForMinute2 + rowsForMinute3);
+      verifyCompactedIntervals(minute1, minute2, minute3);
+
+
+      // Use ALL segment granularity for compaction and run concurrent streaming ingestion
+      submitAndVerifyCompactionConfig(Granularities.DAY);
+
+      for (int i = 0; i < 5; i++) {
+        rowsForMinute1 += generateData(minute1, streamEventWriter);
+        rowsForMinute2 += generateData(minute2, streamEventWriter);
+        rowsForMinute3 += generateData(minute3, streamEventWriter);
+
+        //ensureNoLockContention();
+        compactionResource.forceTriggerAutoCompaction();
+        ensureRowCount(rowsForMinute1 + rowsForMinute2 + rowsForMinute3);
+        printTaskStatuses();
+        checkAndSetConcurrentLocks();
+      }
+
+      // Verify the state with all granularity
+      ensureSegmentsCount(1);
+      ensureSegmentsLoaded();
+      ensureRowCount(rowsForMinute1 + rowsForMinute2 + rowsForMinute3);
+      verifyCompactedIntervals(Intervals.of("2000-01-01/2000-01-02"));
+      printTaskStatuses();
+
+
+      // Concurrent locks should ensure that there are no failed tasks
+      Assert.assertEquals(0, getFailedTaskCount());
+      Assert.assertTrue(concurrentAppendAndReplaceLocksExisted);
+    }
+  }
+
+  /**
+   * Retries until the segment count is as expected.
+   */
+  private void ensureSegmentsCount(int numExpectedSegments)
```

Comment:
It seems a lot of the private methods were copied over from other ITs.
Is it possible reuse this code instead of copying it over?